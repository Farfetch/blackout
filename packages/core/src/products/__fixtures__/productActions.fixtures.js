// @TODO: remove this file in the next version
export const mockPreferedMerchantId = 789;
export const mockSize = {
  id: 25,
  scale: '206',
  stock: [{ merchantId: 10937, quantity: 2 }],
};
export const mockSizePreferedMerchant = {
  id: 26,
  scale: '206',
  stock: [{ merchantId: mockPreferedMerchantId, quantity: 2 }],
};
export const mockProduct = {
  id: 123,
  preferedMerchant: {
    merchantId: mockPreferedMerchantId,
  },
  sizes: [
    mockSize,
    {
      id: 1,
      scale: 1,
      stock: [
        { merchantId: 111, quantity: 2 },
        { merchantId: 222, quantity: 1 },
      ],
    },
    {
      id: 2,
      scale: 2,
      stock: [{ merchantId: 111, quantity: 2 }],
    },
    {
      id: 3,
      scale: 3,
      stock: [{ merchantId: 333, quantity: 2 }],
    },
  ],
};
export const mockQuantity = 2;
export const mockData = {
  product: mockProduct,
  size: mockSize,
  quantity: mockQuantity,
};

export const mockBagItem = {
  id: 1,
  quantity: 1,
  merchant: 111,
  size: {
    id: 2,
    scale: 2,
    stock: [{ merchantId: 111, quantity: 2 }],
  },
  product: mockProduct,
};
