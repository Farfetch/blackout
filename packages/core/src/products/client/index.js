import { warnDeprecatedMethod } from '../../helpers';

warnDeprecatedMethod(
  '@farfetch/blackout-core',
  '@farfetch/blackout-core/products/client',
  '@farfetch/blackout-core/products',
);

/**
 * Products clients.
 *
 * @module products/client
 * @category Products
 * @subcategory Clients
 */
export { default as getSet } from './getSet';
export { default as getProductFittings } from './getProductFittings';
