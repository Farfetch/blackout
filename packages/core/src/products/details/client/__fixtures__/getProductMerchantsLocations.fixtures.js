import join from 'proper-url-join';
import moxios from 'moxios';

/**
 * Response payloads.
 */
export default {
  success: params => {
    moxios.stubRequest(
      join(
        '/api/commerce/v1/products',
        params.productId,
        'variants',
        params.variantId,
        'merchantsLocations',
      ),
      {
        method: 'get',
        response: params.response,
        status: 200,
      },
    );
  },
  failure: params => {
    moxios.stubRequest(
      join(
        '/api/commerce/v1/products',
        params.productId,
        'variants',
        params.variantId,
        'merchantsLocations',
      ),
      {
        method: 'get',
        response: 'stub error',
        status: 404,
      },
    );
  },
};
