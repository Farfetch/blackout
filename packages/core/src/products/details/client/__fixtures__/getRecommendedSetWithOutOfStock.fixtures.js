import join from 'proper-url-join';
import moxios from 'moxios';

/**
 * Response payloads.
 */
export default {
  success: params => {
    moxios.stubRequest(join('/api/commerce/v1/recommendedsets', params.id), {
      response: params.response,
      status: 200,
    });
  },
  failure: params => {
    moxios.stubRequest(join('/api/commerce/v1/recommendedsets', params.id), {
      response: 'stub error',
      status: 404,
    });
  },
};
