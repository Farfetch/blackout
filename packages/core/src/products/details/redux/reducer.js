/**
 * @module products/details/reducer
 * @category Products details
 * @subcategory Reducer
 */
import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createMergedObject } from '../../../helpers/redux';

export const INITIAL_STATE = {
  attributes: {
    error: {},
    isLoading: {},
  },
  colorGrouping: {
    currentPageIndex: {},
    error: {},
    isLoading: {},
  },
  grouping: {
    currentPageIndex: {},
    error: {},
    isLoading: {},
  },
  groupingProperties: {
    error: {},
    isLoading: {},
  },
  error: {},
  fittings: {
    error: {},
    isLoading: {},
  },
  id: null,
  isHydrated: {},
  isLoading: {},
  measurements: {
    error: {},
    isLoading: {},
  },
  merchantsLocations: {
    error: {},
    isLoading: {},
  },
  recommendedSets: {
    error: {},
    isLoading: {},
  },
  recommendedSetsWithOutOfStock: {
    error: {},
    isLoading: {},
  },
  sets: {
    error: {},
    isLoading: {},
  },
  sizeguides: {
    error: {},
    isLoading: {},
  },
  sizes: {
    error: {},
    isLoading: {},
  },
  sizeScale: {
    error: null,
    isLoading: {},
  },
};

const error = (
  state = INITIAL_STATE.error,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_DETAILS_REQUEST:
      return {
        ...state,
        [action.payload.productId]: undefined,
      };
    case actionTypes.GET_PRODUCT_DETAILS_FAILURE:
      return {
        ...state,
        [action.payload.productId]: action.payload.error,
      };
    default:
      return state;
  }
};

const id = (
  state = INITIAL_STATE.id,
  /* istanbul ignore next */ action = {},
) => {
  if (action.type === actionTypes.GET_PRODUCT_DETAILS_SUCCESS) {
    return action.payload.productId;
  }
  return state;
};

const isHydrated = (
  state = INITIAL_STATE.isHydrated,
  /* istanbul ignore next */ action = {},
) => {
  if (action.type === actionTypes.DEHYDRATE_PRODUCT_DETAILS) {
    return {
      ...state,
      [action.payload.productId]: false,
    };
  }
  return state;
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_DETAILS_REQUEST:
      return {
        ...state,
        [action.payload.productId]: true,
      };
    case actionTypes.GET_PRODUCT_DETAILS_SUCCESS:
      return { ...state, [action.payload.productId]: false };
    case actionTypes.GET_PRODUCT_DETAILS_FAILURE:
      return { ...state, [action.payload.productId]: undefined };
    default:
      return state;
  }
};

const attributes = (
  state = INITIAL_STATE.attributes,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_ATTRIBUTES_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: true,
        },
        error: INITIAL_STATE.attributes.error,
      };
    case actionTypes.GET_PRODUCT_ATTRIBUTES_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: false,
        },
      };
    case actionTypes.GET_PRODUCT_ATTRIBUTES_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const colorGrouping = (
  state = INITIAL_STATE.colorGrouping,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_COLOR_GROUPING_REQUEST:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: true,
        },
        error: INITIAL_STATE.colorGrouping.error,
      };
    case actionTypes.GET_COLOR_GROUPING_SUCCESS:
      return {
        ...state,
        currentPageIndex: {
          ...state.currentPageIndex,
          [action.payload.productId]: action.meta.number,
        },
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: false,
        },
      };
    case actionTypes.GET_COLOR_GROUPING_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const fittings = (
  state = INITIAL_STATE.fittings,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.FETCH_PRODUCT_FITTINGS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: true,
        },
        error: INITIAL_STATE.fittings.error,
      };
    case actionTypes.FETCH_PRODUCT_FITTINGS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: false,
        },
      };
    case actionTypes.FETCH_PRODUCT_FITTINGS_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.meta.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const measurements = (
  state = INITIAL_STATE.measurements,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_MEASUREMENTS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: true,
        },
        error: INITIAL_STATE.measurements.error,
      };
    case actionTypes.GET_MEASUREMENTS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: false,
        },
      };
    case actionTypes.GET_MEASUREMENTS_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const recommendedSetsReducer = (
  state = INITIAL_STATE.recommendedSets,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_RECOMMENDED_SET_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.recommendedSetId]: true,
        },
        error: INITIAL_STATE.recommendedSets.error,
      };
    case actionTypes.GET_RECOMMENDED_SET_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.recommendedSetId]: false,
        },
      };
    case actionTypes.GET_RECOMMENDED_SET_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.recommendedSetId]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.recommendedSetId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const recommendedSetsWithOutOfStock = (
  state = INITIAL_STATE.recommendedSetsWithOutOfStock,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.FETCH_RECOMMENDED_SET_WITH_OOS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.recommendedSetId]: true,
        },
        error: INITIAL_STATE.recommendedSetsWithOutOfStock.error,
      };
    case actionTypes.FETCH_RECOMMENDED_SET_WITH_OOS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.recommendedSetId]: false,
        },
      };
    case actionTypes.FETCH_RECOMMENDED_SET_WITH_OOS_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.recommendedSetId]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.recommendedSetId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const sets = (
  state = INITIAL_STATE.sets,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_SET_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.setId]: true,
        },
        error: INITIAL_STATE.sets.error,
      };
    case actionTypes.GET_SET_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.setId]: false,
        },
      };
    case actionTypes.GET_SET_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.setId]: undefined,
        },
        error: {
          ...state.error,
          [action.meta.setId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const sizeguides = (
  state = INITIAL_STATE.sizeguides,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_SIZEGUIDES_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: true,
        },
        error: INITIAL_STATE.sizeguides.error,
      };
    case actionTypes.GET_PRODUCT_SIZEGUIDES_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: false,
        },
      };
    case actionTypes.GET_PRODUCT_SIZEGUIDES_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.meta.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const sizes = (
  state = INITIAL_STATE.sizes,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_SIZES_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: true,
        },
        error: INITIAL_STATE.sizes.error,
      };
    case actionTypes.GET_PRODUCT_SIZES_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: false,
        },
      };
    case actionTypes.GET_PRODUCT_SIZES_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.meta.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

// @TODO: Remove this reducer in version 2.0.0.
const sizeScale = (
  state = INITIAL_STATE.sizeScale,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_SIZESCALE_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.scaleId]: true,
        },
        error: INITIAL_STATE.sizeScale.error,
      };
    case actionTypes.GET_SIZESCALE_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.scaleId]: false,
        },
      };
    case actionTypes.GET_SIZESCALE_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.scaleId]: undefined,
        },
        error: {
          scaleId: action.payload.scaleId,
          message: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const merchantsLocations = (
  state = INITIAL_STATE.merchantsLocations,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_MERCHANTS_LOCATIONS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: true,
        },
        error: INITIAL_STATE.merchantsLocations.error,
      };
    case actionTypes.GET_PRODUCT_MERCHANTS_LOCATIONS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: false,
        },
      };
    case actionTypes.GET_PRODUCT_MERCHANTS_LOCATIONS_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.meta.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const grouping = (
  state = INITIAL_STATE.grouping,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_GROUPING_REQUEST:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: true,
        },
        error: INITIAL_STATE.grouping.error,
      };
    case actionTypes.GET_PRODUCT_GROUPING_SUCCESS:
      return {
        ...state,
        currentPageIndex: {
          ...state.currentPageIndex,
          [action.payload.productId]: action.meta.number,
        },
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: false,
        },
      };
    case actionTypes.GET_PRODUCT_GROUPING_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const groupingProperties = (
  state = INITIAL_STATE.groupingProperties,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCT_GROUPING_PROPERTIES_REQUEST:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: true,
        },
        error: INITIAL_STATE.groupingProperties.error,
      };
    case actionTypes.GET_PRODUCT_GROUPING_PROPERTIES_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: false,
        },
      };
    case actionTypes.GET_PRODUCT_GROUPING_PROPERTIES_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.productId]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.productId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const entitiesMapper = {
  // The motivation for this was some data mismatch:
  // the `/api/products{id}/variantsMeasurements`
  // endpoint returns something different
  // from the details result (which is what we want).
  // However, if the response is an empty array,
  // `normalizr` seems to do nothing to the store.
  // I think this issue:
  // https://github.com/paularmstrong/normalizr/issues/290
  // portraits the problem as well.
  [actionTypes.GET_MEASUREMENTS_SUCCESS]: (
    state,
    { payload: { entities, productId } },
  ) => {
    const newMeasurements = entities.products[productId].measurements;
    const newState = createMergedObject(state, entities);

    newState.products[productId].measurements = newMeasurements;

    return newState;
  },
  [actionTypes.RESET_DETAILS_ENTITIES]: state => {
    const {
      products,
      recommendedSets,
      recommendedSetsWithOutOfStock,
      sets,
      ...rest
    } = state;

    return rest;
  },
};

export const getError = state => state.error;
export const getId = state => state.id;
export const getIsHydrated = state => state.isHydrated;
export const getIsLoading = state => state.isLoading;
// Attributes
export const getAttributesError = state => state.attributes.error;
export const getAreAttributesLoading = state => state.attributes.isLoading;
// Color grouping
export const getColorGroupingError = state => state.colorGrouping.error;
export const getIsColorGroupingLoading = state => state.colorGrouping.isLoading;
export const getColorGroupingCurrentPageIndex = state =>
  state.colorGrouping.currentPageIndex;
// Grouping
export const getGroupingError = state => state.grouping.error;
export const getIsGroupingLoading = state => state.grouping.isLoading;
export const getGroupingCurrentPageIndex = state =>
  state.grouping.currentPageIndex;
// Grouping Properties
export const getGroupingPropertiesError = state =>
  state.groupingProperties.error;
export const getIsGroupingPropertiesLoading = state =>
  state.groupingProperties.isLoading;
// Fittings
export const getFittingsError = state => state.fittings.error;
export const getAreFittingsLoading = state => state.fittings.isLoading;
// Measurements
export const getMeasurementsError = state => state.measurements.error;
export const getAreMeasurementsLoading = state => state.measurements.isLoading;
// Recommended sets
export const getRecommendedSetError = state => state.recommendedSets.error;
export const getRecommendedSetIsLoading = state =>
  state.recommendedSets.isLoading;
// Recommended sets with out of stock
export const getRecommendedSetWithOutOfStockError = state =>
  state.recommendedSetsWithOutOfStock.error;
export const getRecommendedSetWithOutOfStockIsLoading = state =>
  state.recommendedSetsWithOutOfStock.isLoading;
// Sets
export const getSetError = state => state.sets.error;
export const getSetIsLoading = state => state.sets.isLoading;
// Sizeguides
export const getSizeguidesError = state => state.sizeguides.error;
export const getAreSizeguidesLoading = state => state.sizeguides.isLoading;
// Sizes
export const getSizesError = state => state.sizes.error;
export const getAreSizesLoading = state => state.sizes.isLoading;
// Sizescale
export const getSizeScaleError = state => state.sizeScale.error;
export const getSizeScaleIsLoading = state => state.sizeScale.isLoading;
// Merchants Locations
export const getMerchantsLocationsError = state =>
  state.merchantsLocations.error;
export const getMerchantsLocationsIsLoading = state =>
  state.merchantsLocations.isLoading;

const reducers = combineReducers({
  attributes,
  colorGrouping,
  error,
  fittings,
  grouping,
  groupingProperties,
  id,
  isHydrated,
  isLoading,
  measurements,
  merchantsLocations,
  recommendedSets: recommendedSetsReducer,
  recommendedSetsWithOutOfStock,
  sets,
  sizeguides,
  sizes,
  sizeScale,
});

/**
 * Reducer for products details state.
 *
 * @function productsDetailsReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_DETAILS_STATE) {
    return reducers(INITIAL_STATE, action);
  }

  return reducers(state, action);
};
