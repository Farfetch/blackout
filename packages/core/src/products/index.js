// @TODO: remove this file in the next version
export { default as withProductActions } from './withProductActions';
