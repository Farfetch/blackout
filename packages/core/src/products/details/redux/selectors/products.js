import {
  createGetItemInBag,
  getItemWholeQuantity,
} from '../../../../bags/redux';
import { getError, getIsHydrated, getIsLoading } from '../reducer';
import { getProduct } from '../../../../entities/redux/selectors';
import get from 'lodash/get';

/**
 * Returns the error given by product actions.
 *
 * @function
 * @memberof module:products/details/selectors
 *
 * @param {object} state - Application state.
 * @param {number} id - Product id.
 *
 * @returns {object} Product details error.
 */
export const getProductError = (state, id) => getError(state.details)[id];

/**
 * Returns the hydrated condition from product details.
 *
 * @function
 * @memberof module:products/details/selectors
 *
 * @param {object} state - Application state.
 * @param {number} id - Product id.
 *
 * @returns {boolean} If a certain product is hydrated or not.
 */
export const isProductHydrated = (state, id) =>
  getIsHydrated(state.details)[id];

/**
 * Returns the loading condition from product details.
 *
 * @function
 * @memberof module:products/details/selectors
 *
 * @param {object} state - Application state.
 * @param {number} id - Product id.
 *
 * @returns {boolean} If a certain product is loading or not.
 */
export const isProductLoading = (state, id) => getIsLoading(state.details)[id];

/**
 * Returns the fetched status of a specific product.
 *
 * @function
 * @memberof module:products/details/selectors
 *
 * @param {object} state - Application state.
 * @param {number} id - Product id.
 *
 * @returns {boolean} If a certain product has been fetched or not.
 */
export const isProductFetched = (state, id) =>
  (getIsLoading(state.details).hasOwnProperty(id) ||
    getIsHydrated(state.details).hasOwnProperty(id)) &&
  isProductLoading(state, id) === false;

/**
 * Returns if the product is duplicated or not.
 *
 * @function
 * @memberof module:products/details/selectors
 *
 * @param {object} state - Application state.
 * @param {number} id - Product id.
 *
 * @returns {boolean|undefined} If a certain product is duplicated or not.
 */
export const isProductDuplicated = (state, id) => {
  const product = getProduct(state, id);

  return get(product, 'isDuplicated');
};

/**
 * Returns all the info about breadcrumbs at PDP.
 *
 * @function
 * @memberof module:products/details/selectors
 *
 * @param {object} state - Application state.
 * @param {number} id - Product id.
 *
 * @returns {object|undefined} Breadcrumbs info.
 */
export const getProductBreadcrumbs = (state, id) => {
  const product = getProduct(state, id);

  return get(product, 'breadCrumbs');
};

/**
 * Creates a function responsible for checking the remaining available quantity
 * of a product of a given size, based on its quantity in the bag.
 * This creator should only be used in cases/projects you have the need
 * to have product details and bag logics/reducer at the same time.
 *
 * @function
 * @memberof module:products/details/selectors
 *
 * @param {object} state - Application state.
 * @param {number} productId - Numeric identifier of the product.
 *
 * @returns {Function} Function that returns a Number,
 * which is the difference between the total quantity of product size and the
 * respective bag quantity.
 *
 * @example
 * import { createGetProductRemainingQuantity } from '@farfetch/blackout-core/products/details/redux';
 *
 * const mapStateToProps = (state, { productId }) => ({
 *     getRemainingQuantity: createGetProductRemainingQuantity(state, productId)
 * });
 *
 * render(
 *  const {getRemainingQuantity, sizes } = this.props;
 *  const options = sizes.map(size => {
 *      const remainingQuantity = getRemainingQuantity(size.id);
 *      ...
 *  });
 * )
 */
export const createGetProductRemainingQuantity =
  (state, productId) => sizeId => {
    const getItemInBag = createGetItemInBag(state);
    const product = getProduct(state, productId);
    const size = product.sizes.find(({ id }) => id === sizeId);
    const stockAvailable = get(size, 'globalQuantity', 0);
    const bagItem = getItemInBag({ product, size });

    if (bagItem) {
      const itemWholeQuantity = getItemWholeQuantity(state, bagItem);

      return stockAvailable - itemWholeQuantity;
    }

    return stockAvailable;
  };
