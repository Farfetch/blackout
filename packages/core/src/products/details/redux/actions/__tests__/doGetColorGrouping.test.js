import { doGetColorGrouping } from '../';
import { mockColorGrouping, mockProductId } from 'tests/__fixtures__/products';
import { mockStore } from '../../../../../../tests';
import find from 'lodash/find';
import reducer, { actionTypes } from '../../';

const productDetailsMockStore = (state = {}) =>
  mockStore({ details: reducer() }, state);

describe('doGetColorGrouping() action creator', () => {
  let store;
  const getColorGrouping = jest.fn();
  const action = doGetColorGrouping(getColorGrouping);
  const expectedConfig = undefined;
  const normalizeSpy = jest.spyOn(require('normalizr'), 'normalize');

  beforeEach(() => {
    jest.clearAllMocks();

    store = productDetailsMockStore({
      entities: {
        products: {
          [mockProductId]: {
            id: mockProductId,
            colorGrouping: [mockColorGrouping[0]],
          },
        },
      },
    });
  });

  it('should create the correct actions for when the get color grouping procedure fails', async () => {
    const expectedError = new Error('Get color grouping error');

    getColorGrouping.mockRejectedValueOnce(expectedError);

    expect.assertions(4);

    try {
      await store.dispatch(action(mockProductId));
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(getColorGrouping).toHaveBeenCalledTimes(1);
      expect(getColorGrouping).toHaveBeenCalledWith(
        mockProductId,
        {},
        expectedConfig,
      );
      expect(store.getActions()).toEqual([
        {
          payload: { productId: mockProductId },
          type: actionTypes.GET_COLOR_GROUPING_REQUEST,
        },
        {
          payload: { error: expectedError, productId: mockProductId },
          type: actionTypes.GET_COLOR_GROUPING_FAILURE,
        },
      ]);
    }
  });

  it('should create the correct actions for when the get color grouping procedure is successful', async () => {
    getColorGrouping.mockResolvedValueOnce(mockColorGrouping[1]);

    const query = { pageSize: 10 };

    await store.dispatch(action(mockProductId, query));

    const actionResults = store.getActions();

    expect(normalizeSpy).toHaveBeenCalledTimes(1);
    expect(getColorGrouping).toHaveBeenCalledTimes(1);
    expect(getColorGrouping).toHaveBeenCalledWith(
      mockProductId,
      query,
      expectedConfig,
    );
    expect(actionResults).toEqual([
      {
        payload: { productId: mockProductId },
        type: actionTypes.GET_COLOR_GROUPING_REQUEST,
      },
      expect.objectContaining({
        payload: expect.any(Object),
        type: actionTypes.GET_COLOR_GROUPING_SUCCESS,
      }),
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.GET_COLOR_GROUPING_SUCCESS,
      }),
    ).toMatchSnapshot('Get color grouping success payload');
  });

  it('should create the correct actions for when the get color grouping procedure is successful with no previous data', async () => {
    // Store without a previous color grouping entry
    store = productDetailsMockStore({
      entities: {
        products: {
          [mockProductId]: {
            id: mockProductId,
          },
        },
      },
    });

    getColorGrouping.mockResolvedValueOnce(mockColorGrouping[1]);

    await store.dispatch(action(mockProductId));

    const actionResults = store.getActions();

    expect(normalizeSpy).toHaveBeenCalledTimes(1);
    expect(getColorGrouping).toHaveBeenCalledTimes(1);
    expect(getColorGrouping).toHaveBeenCalledWith(
      mockProductId,
      {},
      expectedConfig,
    );
    // Just ensure that the success happens
    expect(actionResults).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          payload: expect.any(Object),
          type: actionTypes.GET_COLOR_GROUPING_SUCCESS,
        }),
      ]),
    );
  });
});
