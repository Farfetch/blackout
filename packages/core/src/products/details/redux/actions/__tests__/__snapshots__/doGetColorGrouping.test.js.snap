// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doGetColorGrouping() action creator should create the correct actions for when the get color grouping procedure is successful: Get color grouping success payload 1`] = `
Object {
  "meta": Object {
    "number": 1,
  },
  "payload": Object {
    "entities": Object {
      "products": Object {
        "11766695": Object {
          "DEPRECATED_images": undefined,
          "colorGrouping": Array [
            Object {
              "entries": Array [
                Object {
                  "color": "9075 WHITE",
                  "digitalAssets": Array [
                    Object {
                      "displayOrder": 1,
                      "mediaType": "image/jpeg",
                      "order": 1,
                      "size": "54",
                      "sources": Object {
                        "54": "https://api.blackout.com/14/26/08/35/14260835_20007113_54.jpg",
                      },
                      "type": 0,
                      "url": "https://api.blackout.com/14/26/08/35/14260835_20007113_54.jpg",
                    },
                  ],
                  "hasStock": true,
                  "id": 14260832,
                  "isDefault": true,
                  "order": 1,
                  "slug": "woman-scarf-14260835",
                  "variantId": null,
                },
                Object {
                  "color": "9075 PINK",
                  "digitalAssets": Array [
                    Object {
                      "displayOrder": 1,
                      "mediaType": "image/jpeg",
                      "order": 1,
                      "size": "54",
                      "sources": Object {
                        "54": "https://api.blackout.com/14/26/08/34/14260834_20007111_54.jpg",
                      },
                      "type": 0,
                      "url": "https://api.blackout.com/14/26/08/34/14260834_20007111_54.jpg",
                    },
                  ],
                  "hasStock": true,
                  "id": 14260833,
                  "isDefault": null,
                  "order": 1,
                  "slug": "woman-scarf-14260834",
                  "variantId": null,
                },
              ],
              "number": 0,
              "totalItems": 2,
              "totalPages": 1,
            },
            Object {
              "entries": Array [
                Object {
                  "color": "9075 YELLOW",
                  "digitalAssets": Array [
                    Object {
                      "displayOrder": 1,
                      "mediaType": "image/jpeg",
                      "order": 1,
                      "size": "54",
                      "sources": Object {
                        "54": "https://api.blackout.com/14/26/08/35/14260835_20007113_54.jpg",
                      },
                      "type": 0,
                      "url": "https://api.blackout.com/14/26/08/35/14260835_20007113_54.jpg",
                    },
                  ],
                  "hasStock": true,
                  "id": 14260834,
                  "isDefault": true,
                  "order": 1,
                  "slug": "woman-scarf-14260835",
                  "variantId": null,
                },
                Object {
                  "color": "9075 BLACK",
                  "digitalAssets": Array [
                    Object {
                      "displayOrder": 1,
                      "mediaType": "image/jpeg",
                      "order": 1,
                      "size": "54",
                      "sources": Object {
                        "54": "https://api.blackout.com/14/26/08/34/14260834_20007111_54.jpg",
                      },
                      "type": 0,
                      "url": "https://api.blackout.com/14/26/08/34/14260834_20007111_54.jpg",
                    },
                  ],
                  "hasStock": true,
                  "id": 14260835,
                  "isDefault": null,
                  "order": 1,
                  "slug": "woman-scarf-14260834",
                  "variantId": null,
                },
              ],
              "number": 1,
              "totalItems": 2,
              "totalPages": 1,
            },
          ],
          "customAttributes": undefined,
          "groupedEntries": undefined,
          "grouping": undefined,
          "groupingProperties": undefined,
          "id": 11766695,
          "images": undefined,
          "merchant": undefined,
          "price": undefined,
          "prices": undefined,
          "sizes": undefined,
          "tag": Object {
            "id": undefined,
            "name": undefined,
          },
          "variants": undefined,
        },
      },
    },
    "productId": 11766695,
    "result": 11766695,
  },
  "type": "@farfetch/blackout-core/GET_COLOR_GROUPING_SUCCESS",
}
`;
